File: 2023/14.py
from copy import deepcopy

input = open(f"{__file__.replace('py', 'input')}").read().strip().split("\n")
input = [[*x] for x in input]


def tilt(platform):
    for i in range(len(platform)):
        for j in range(len(platform[i])):
            if platform[i][j] != "O":
                continue

            k = i - 1
            while True:
                if k == -1 or platform[k][j] == "#" or platform[k][j] == "O":
                    break
                k -= 1
            platform[i][j], platform[k + 1][j] = platform[k + 1][j], platform[i][j]
    return platform


def load(platform):
    total_sum = 0
    for i in range(len(platform)):
        total_sum += (len(platform) - i) * platform[i].count("O")
    return total_sum


print(load(tilt(deepcopy(input))))

cache = {}
platform = input
t = 0
while t < 1_000_000_000:
    if repr(platform) in cache:
        t = 1_000_000_000 - ((1_000_000_000 - t) % (t - cache[repr(platform)]))
    cache[repr(platform)] = t
    t += 1

    platform = [list(r) for r in zip(*tilt(platform)[::-1])]
    platform = [list(r) for r in zip(*tilt(platform)[::-1])]
    platform = [list(r) for r in zip(*tilt(platform)[::-1])]
    platform = [list(r) for r in zip(*tilt(platform)[::-1])]

print(load(platform))

File: 2023/14.rb
input = File.read(File.basename(__FILE__).gsub('rb', 'input')).strip.split("\n").map(&:chars)

def tilt(platform)
  (0...platform.size).each do |i|
    (0...platform[i].size).each do |j|
      next unless platform[i][j] == 'O'

      k = i - 1
      loop do
        break if k == -1 || platform[k][j] == '#' || platform[k][j] == 'O'

        k -= 1
      end
      platform[i][j], platform[k + 1][j] = platform[k + 1][j], platform[i][j]
    end
  end
  platform
end

def load(platform)
  platform.reverse.map.with_index { (_2 + 1) * _1.count('O') }.sum
end

p load(tilt(input.map(&:clone)))

@cache = {}
platform = input.map(&:clone)
t = 0
while t < 1_000_000_000
  if @cache[platform.map(&:join).join]
    t = 1_000_000_000 - ((1_000_000_000 - t) % (t - @cache[platform.map(&:join).join]))
  end
  @cache[platform.map(&:join).join] = t
  t += 1

  platform = tilt(platform).transpose.map(&:reverse)
  platform = tilt(platform).transpose.map(&:reverse)
  platform = tilt(platform).transpose.map(&:reverse)
  platform = tilt(platform).transpose.map(&:reverse)
end
p load(platform)
